Prepare release notes for Qiskit Terra 0.19

jakelishman opened the pull request:
### Summary

The main object of this PR is to move the release notes for 0.19 into a separate folder, and to reformat them into a consistent style with correct formatting and working Sphinx cross-references.

In doing so, I also came across some places with missing documentation, and a couple of places where the Sphinx documentation was being generated twice for quite a few objects, which messed with the ability to cross-reference correctly.  This PR is then actually in a few commits, which may be viewed separately to see the changes more easily:

- move the release notes into `releasenotes/notes/0.19`
- add objects that were missing from the API docs/module imports completely (that looked like they _should_ have been included - I didn't do it indiscriminately)
- updated a couple of deprecation notices that incorrectly said 0.18 due to their milestone getting moved
- reorganised the `qiskit.pulse` documentation to avoid duplicating definitions of all its objects (this could be split into a separate PR)
- actually reformatted the release notes

### Details and comments

Most of the release notes changes are mechanical - fixing typos, formatting and cross-reference links.  In a few cases, there was insufficient context in the release note itself when read in isolation, so those I mostly rewrote.  If you want to review the changes only, you may want to filter the diff only to the single commit named "Reformat 0.19 release notes" (its hash may change if I rebase).

The main documentation _addition_ was hooking the approximate quantum compiler into the actual tree.  This then necessitated a couple of fixes in its documentation to account for Sphinx errors that had previously gone unnoticed.  Adding a few things into the import path created some cyclical imports, so those are moved into runtime imports.

While the changes to the `qiskit.pulse` documentation _look_ big, in reality, the rendered documentation is almost identical (because it's already good - I didn't want to mess with it!).  I just rearranged _how_ it gets generated, and which docstrings are responsible for various parts of it.  The main developer-facing change is that now there is only one canonical reference path for every object in the `qiskit.pulse` namespace.  The builder objects, `Schedule` and `ScheduleBlock` are in the top-level `qiskit.pulse`, and most other stuff is in `qiskit.pulse.instructions`, `.library`, `.transforms`, etc (with no further nesting within these).  This means that suffix references will now generally work (e.g. `` :class:`.Gaussian` `` will now actually resolve, and be equal to `` :class:`~qiskit.pulse.library.Gaussian` ``).  This should also help users searching for documentation, because there won't be duplication in the results.

---

This is likely going to be the last PR for 0.19, and the one that gets tagged as the release.  On hold til all other PRs are in, at which point I'll rebase it and write a release prelude.


Todo before merge:
- [x] merge in remaining 0.19 PRs and move release notes
- [x] rewrite QASM 3 and control-flow release notes to reference the builder interface/promote control-flow more

mtreinish commented on qiskit/transpiler/passes/optimization/echo_rzx_weyl_decomposition.py:
```diff
@@ -38,7 +35,7 @@ class EchoRZXWeylDecomposition(TransformationPass):
     """
 
-    def __init__(self, backend: basebackend):
+    def __init__(self, backend: BaseBackend):
```
This has the same bug as the other pass, I guess I'll bundle this as one more critical bugfix PR before the release....

mtreinish replied:
Resolved with #7331

mtreinish reviewed (commented):
I just went through the doc changes to start here are some quick comments. I'm gonna take a break before starting the release note meat of this PR to start preparing a fix commit for those 2 broken transpiler passes before we're committed to the API in the release.

Cryoris commented on qiskit/circuit/library/arithmetic/exact_reciprocal.py:
```diff
@@ -31,14 +31,16 @@ def __init__(
             num_state_qubits: The number of qubits representing the value to invert.
-            scaling: Scaling factor of the reciprocal function, i.e. to compute
-             ::math:: `scaling / x`.
+            scaling: Scaling factor of the reciprocal function, i.e. to compute :math:`scaling / x`.
```
This is probably gonna be rendered poorly, how about
```suggestion
            scaling: Scaling factor :math:`s` of the reciprocal function, i.e. to compute
                :math:`s / x`.

```

jakelishman replied:
Done in aa157bc.

Cryoris commented on qiskit/circuit/library/arithmetic/exact_reciprocal.py:
```diff
@@ -31,14 +31,16 @@ def __init__(
-            scaling: Scaling factor of the reciprocal function, i.e. to compute
-             ::math:: `scaling / x`.
+            scaling: Scaling factor of the reciprocal function, i.e. to compute :math:`scaling / x`.
             neg_vals: Whether x might represent negative values. In this case the first qubit is
```
```suggestion
            neg_vals: Whether :math:`x` might represent negative values. In this case the first qubit is
```

jakelishman replied:
Done in aa157bc.

Cryoris commented on qiskit/circuit/library/arithmetic/exact_reciprocal.py:
```diff
@@ -31,14 +31,16 @@ def __init__(
+
+        .. note::
+
             It is assumed that the binary string x represents a number < 1.
```
```suggestion
            It is assumed that the binary string :math:`x` represents a number :math:`< 1`.
```

jakelishman replied:
Done in aa157bc.

jakelishman commented on qiskit/pulse/__init__.py:
```diff
@@ -10,101 +10,38 @@
+.. automodule:: qiskit.pulse.transforms
+.. automodule:: qiskit.pulse.builder
 
+.. currentmodule:: qiskit.pulse
```
`automodule` resets the `currentmodule`, so this restores it.  In practice, `qiskit.pulse.builder` also sets the `currentmodule` to `qiskit.pulse`, but I didn't want to be order-dependent if things like that change.

jakelishman commented on qiskit/pulse/instructions/__init__.py:
```diff
@@ -10,7 +10,14 @@
+Instructions (:mod:`qiskit.pulse.instructions`)
+===============================================
+
+The ``instructions`` module holds the various :obj:`Instruction`\ s which are supported by
```
Without a preceding `.`, Sphinx tries to match names within the `currentmodule` (`qiskit.pulse.instructions` here), then the parent module, and so on.  So it'll find `qiskit.pulse.instruction.Instruction` here, because `Instruction` is done by `automodule` later on here.  A lot of the problem we generally have is that our documentation generation is very implicit via autosummary cascading through a bunch of docstrings.  It's not clear what namespace something is defined in, much of the time...  Here it's much clearer what we mean, because there's an explicit `autoclass`.

jakelishman commented on qiskit/quantum_info/synthesis/xx_decompose/decomposer.py:
```diff
@@ -72,8 +68,9 @@ class XXDecomposer:
 
-    NOTE: If embodiments is not passed, or if an entry is missing, it will be populated as needed
-        using the method _default_embodiment.
+    .. note::
+        If ``embodiments`` is not passed, or if an entry is missing, it will be populated as needed
```
`.. note::` doesn't take any arguments, so it still renders correctly if it doesn't have the line break.  I do usually put them in, though, just for consistency.

jakelishman commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -25,7 +25,7 @@
     """Product formula base class for the decomposition of non-commuting operator exponentials.
 
-    Lie-Trotter and Suzuki inherit this class.
+    :obj:`.LieTrotter` and :obj:`.SuzukiTrotter` inherit from this class.
```
They affect how Sphinx searches for objects.  The current Sphinx module of this (`ProductFormula`) object is `qiskit.synthesis` I think, so if I just wrote `` `:obj:`LieTrotter` ``, it would try in order:
- `LieTrotter`
- `qiskit.synthesis.LieTrotter`
- `qiskit.synthesis.ProductFormula.LieTrotter`

and if none of those matched, it'd fail.  If you put the dot, it searches those same things (but in reverse order), but if it fails, then it searches for _any_ suffix match, so it'd also try things like `qiskit.quantum_info.LieTrotter`, which otherwise wouldn't be in the path at all.  In this particular case, either would work, but I typically use the `.` to trigger the full suffix match - the way `automodule` and our docstrings manage the `currentmodule` is very confusing (hence we have so many broken cross-references even when they try to use the "full" form), and it's much easier just to force a search if you know there's only one valid object.

Cryoris replied:
Ah ok, makes sense -- thanks!

jakelishman replied:
I'm not sure _anything_ about Sphinx' process makes sense, to be honest, but I'm fairly sure this is the way things are!

jakelishman commented on qiskit/synthesis/evolution/suzuki_trotter.py:
```diff
@@ -56,6 +55,18 @@ def __init__(
             Callable[[Union[Pauli, SparsePauliOp], float], QuantumCircuit]
         ] = None,
     ) -> None:
+        """
```
For some reason, yes - I think it might be a Sphinx parsing bug.  If I let Sphinx try to inherit the `__init__` docstring from the base class, it complains about invalid indentation (perhaps it doesn't join the indentation levels together correctly or something).

jakelishman commented on docs/apidocs/transpiler_builtin_plugins.rst:
```diff
@@ -0,0 +1,10 @@
+.. toctree::
+   :maxdepth: 2
+
+   synthesis_aqc
```
Yeah, I didn't put it in here because it didn't have any user-facing documentation to link to.  If I make a stub file, it's basically just going to be the text "The default unitary synthesis plugin".

jakelishman commented:
I'm going to leave the changes for now, just to ease the CI - I've got them done locally, but I'll not push til more stuff has merged.

jakelishman commented:
Well, I was going to leave things, then I accidentally closed the PR rather than just commenting, and re-opening re-triggered the build anyway.  So now they're done.

mtreinish commented on releasenotes/notes/0.19/add-qubit-subset-to-bip-mapper-e1c6234d04484d58.yaml:
```diff
@@ -2,7 +2,7 @@
   - |
     Introduced a new option ``qubit_subset`` to the constructor of
-    :class:`qiskit.transpiler.passes.BIPMapping`.
+    :class:`.BIPMapping`.
```
Is there any difference with this?

jakelishman replied:
It'll only display `BIPMapping` rather than `qiskit.transpiler.passes.BIPMapping`, but I don't know why I didn't just prefix it with a `~`.  In general, I shortened the output of everything down to the necessary information (mostly `Class.method` and occasionally `module.function` or `module.Class`), because that makes it easier to read and you can always click the link to get the full path if you need it.

mtreinish commented on releasenotes/notes/0.19/add-sparsepauliop-sum-d55fc817c9fded82.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Added :meth:`.SparsePauliOp.sum` as a specialized version of the builtin
+    function ``sum`` for a list of :class:`~qiskit.quantum_info.SparsePauliOp`.
+    This method has better scaling than adding the instances together in a loop.
```
It's a bit weird reading as everything is called `sum` maybe add an example of the builtin.

```suggestion
    Added the :meth:`.SparsePauliOp.sum` method as a specialized version of the builtin
    function ``sum`` (e.g. `sum([x, y])` ) for a list of :class:`~qiskit.quantum_info.SparsePauliOp`.
    This method has better scaling than adding the instances together in a loop.
```

I think this would also benefit an a/b example of what would have been done before and how the new `sum()` method can be used instead with better scaling.

jakelishman replied:
Sure - in this case it'll display as
> Added `SparsePauliOp.sum` as a specialised version of the builtin function ``sum``, ...

(so the context is there), but I can make it clearer, and the example is good.

mtreinish commented on releasenotes/notes/0.19/add-support-to-disable-amplitude-limit-in-parametric-pulses-ef88b77db8c1b06c.yaml:
```diff
@@ -0,0 +1,10 @@
+    which is the base class of :obj:`.Gaussian`, :obj:`.GaussianSquare`, :obj:`.Drag` and :obj:`.Constant`, to
+    enable or disable the amplitude limit of 1 on the pulse. This
+    option gets propagated to the parent class ``Pulse``. With this option enabled,
+    the ``ParametricPulse`` class and its subclasses do not allow amplitudes larger
+    than 1. By default, the value of ``limit_amplitude`` is true, so that the
```
I assume they raise an exception should we document what that is here?

jakelishman replied:
I rewrote this whole note - the note is pretty misleading.  In reality, the option lets you _disable_ the error (which already existed) on an instance-by-instance basis.

mtreinish commented on releasenotes/notes/0.19/add-support-to-disable-amplitude-limit-in-parametric-pulses-ef88b77db8c1b06c.yaml:
```diff
@@ -0,0 +1,10 @@
+    option gets propagated to the parent class ``Pulse``. With this option enabled,
+    the ``ParametricPulse`` class and its subclasses do not allow amplitudes larger
+    than 1. By default, the value of ``limit_amplitude`` is true, so that the
+    amplitude is constrained to be less than 1. See `#6544 <https://github.com/Qiskit/qiskit-terra/issues/6544>`__ for more detail.
```
Should this be an upgrade note as it's potentially a breaking change if someone was constructing a parametric pulse with an amplitude >1 prior to 0.19 it would have worked and now it will error unless they manually change the constructor call to set `limit_amplitude` false

jakelishman replied:
Looking at the history and the code, it's actually that `limit_amplitude` has always existed on the base class as a class attribute, it's just that this PR added an argument so you can _disable_ it on an instance-by-instance basis.  Have rewritten the note to reflect that.

mtreinish commented on releasenotes/notes/0.19/added-multiformat-support-b5d3c7c7c1536951.yaml:
```diff
@@ -0,0 +1,22 @@
+---
+features:
+  - |
+    Using :meth:`.QuantumCircuit.draw` with the ``latex`` drawer will now generate a file
```
We should also mention the standalone drawer function:

```suggestion
    Using :meth:`.QuantumCircuit.draw` or :func:`~qiskit.visualization.circuit_drawer`
    with the ``output`` argument set to ``"latex"`` the drawer will now generate a file
```

mtreinish commented on releasenotes/notes/0.19/added-multiformat-support-b5d3c7c7c1536951.yaml:
```diff
@@ -0,0 +1,22 @@
+    This will save the circuit in the JPEG format (earlier it always used to save it in PNG).
+    Refer to `#6448 <https://github.com/Qiskit/qiskit-terra/issues/6448>`__ for more details.
+
+    Now, if it encounters a filename extension which is not supported, for example::
+      
+      circuit.draw('latex', filename='./file.spooky')
+
+    it will give user a ``ValueError`` to change the filename extension to a supported image format.
+    A similar ``ValueError`` is raised if a user tries to save the output of the ``text`` drawer to a file without using a ``.txt`` extension.
```
I feel like this should also be an upgrade note as it's a potentially breaking change. Especially the text drawer one, I'm tempted to revert that because it seems overly pedantic to me to force people to use a `.txt` extension for a text file.

jakelishman replied:
Yeah, that's exactly what I thought about the `txt` extension when I read this note as well...

jakelishman replied:
Pushed 2c93a9a46 to revert the `txt` change.

mtreinish commented on releasenotes/notes/0.19/added-snapshot-tests-for-backend-mapping-functions-5961300e09f05be0.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Added the parameter  ``filename`` to :func:`~qiskit.visualization.plot_gate_map` and :func:`~qiskit.visualization.plot_coupling_map`, which allows saving the resulting images to a file.
```
While there are no linewrap rules enforced here when we port this over to qiskit/qiskit in the release_notes.rst file a 100 character wide line length is enforced. So it's good to wrap these now so we don't have to manually fix them in the output of `reno report` that we embed in the release notes file.

jakelishman replied:
I reflowed all release notes that looked too long.

mtreinish commented on releasenotes/notes/0.19/bip-improvement-34cd623294a30d83.yaml:
```diff
@@ -1,13 +1,13 @@
-    The ``balanced`` objective function of the ``BIPMapping`` no longer
-    requires a ``BackendProperties``. It will use ``BackendProperties`` if
+    The ``balanced`` objective function of the :obj:`.BIPMapping` no longer
+    requires a :obj:`.BackendProperties`. It will use :obj:`.BackendProperties` if
```
Should we update the earlier releae note that says it requires backend props to be set with 2q error rates in the BackendProperties?

jakelishman replied:
I've combined this release note into the other, and deleted this one.

mtreinish commented on releasenotes/notes/0.19/bip-improvement-34cd623294a30d83.yaml:
```diff
@@ -1,13 +1,13 @@
 upgrade:
   - |
-    The default method for ``BIPMapping`` is now ``balanced`` rather than
+    The default method for :obj:`.BIPMapping` is now ``balanced`` rather than
     ``depth``.
```
I would expand on this a little bit to make it explicit what to do and why this was done.

```suggestion
    The default method for :obj:`.BIPMapping` is now ``balanced`` rather than
    ``depth``. This was done as balanced achieves a generally better result as it
    factors in more than just circuit depth as the objective. If you require the previous
    behavior you should explicitly set the ``method`` kwarg on the :obj:`.BIPMapping`
    to ``"depth"``
```

mtreinish commented on releasenotes/notes/0.19/dag_node_to_op_in_out_node-af2cf9c3e7686285.yaml:
```diff
@@ -0,0 +1,14 @@
+    class and will be used in the future only as the parent class for
+    :class:`~qiskit.dagcircuit.dagnode.DAGOpNode`, :class:`~qiskit.dagcircuit.dagnode.DAGInNode`,
+    and :class:`~qiskit.dagcircuit.dagnode.DAGOutNode`. As part of this deprecation, the
+    following kwargs and associated attributes in :obj:`.DAGNode` are being deprecated: ``type``,
```
```suggestion
    following kwargs and associated attributes in :obj:`.DAGNode` are also being deprecated: ``type``,
```

mtreinish commented on releasenotes/notes/0.19/dag_node_to_op_in_out_node-af2cf9c3e7686285.yaml:
```diff
@@ -0,0 +1,14 @@
+    and :class:`~qiskit.dagcircuit.dagnode.DAGOutNode`. As part of this deprecation, the
+    following kwargs and associated attributes in :obj:`.DAGNode` are being deprecated: ``type``,
+    ``op``, and ``wire``.
+upgrade:
+  - |
+    The previously deprecated ``condition`` kwarg has been removed from
+    :meth:`.DAGCircuit.apply_operation_back` and
+    :meth:`.DAGCircuit.apply_operation_front`.
```
```suggestion
upgrade:
  - |
    The previously deprecated ``condition`` kwarg, which was deprecated as part of the 0.15.0 release,
    has been removed from
    :meth:`.DAGCircuit.apply_operation_back` and
    :meth:`.DAGCircuit.apply_operation_front`. Instead the ``control`` set on the
    :class:`~qiskit.circuit.Instruction` instances being added to the :class:`~qiskit.dagcircuit.DAGCircuit`
    should be used
```

mtreinish reviewed (commented):
I'm about half way through the release notes these are my comments so far. I'd normally wait until I finish reviewing, but I'm spooked now because github refused to let me leave a suggestion as the diff changed and needed to be refreshed and I don't trust it to preserve everything after I do.

mtreinish reviewed (commented):
Also we can delete:

- completemeasfitter-386e94a271ba0180.yaml
-  fix-ScheduleBlock-exceution-012973439c95d8c0.yaml 
-  fix-aarch64-e75601c0b8f2da2c.yaml 
-  fix-assemble-parametric-pulse-eea8fba73c56a69f.yaml 
-  fix-bip-mapper-requirements-930c7dcd5a3b82ba.yaml 
-  fix-complex-param-binding-0fcad20135aa7382.yaml 
- fix-duplicate-mpl-drawer-ipykernel6-c64d9178e31ab9fe.yaml
-  fix-kwargs-mock-f581d5118a1c7589.yaml 
-  fix-qaoa-to-qpy-6364c96671bc8f36.yaml 
-  fix-qnspsa-settings-91d5a9553706cc72.yaml 
-  fix-subgraph-cmap-47021596d94bb39f.yaml 
-  one-gate-identity-optimization-fix-d6d1496d9f5409a6.yaml 
-  single-qubit-conditions-qpy-39a5a9a7204a27dd.yaml 
-  sx-error-map-408f28a6a0c05fd2.yaml 

(if any of those are still present, I saw a few of them as I was reading the notes) as we backported them and they've already been published as release notes. Since we don't really support > 1 stable branch at a time we normally don't want fixes to show up twice

jakelishman commented on releasenotes/notes/0.19/QuantumCircuit.decompose-takes-which-gate-to-decompose-d857da5d0c41fb66.yaml:
```diff
@@ -0,0 +1,17 @@
+  - |
+    :meth:`.QuantumCircuit.decompose` and its corresponding transpiler pass now optionally accept a parameter containing a collection of gate names.
+    If this parameter is given, then only gates with matching names will be decomposed.
+    This supports Unix-shell-style wildcard matches.  For example::
```
My problem here is actually that I couldn't think of any sensible example where you'd want to do this, haha!  The external contributor who wrote this feature used `fnmatch`, whereas I think a much more natural form would have been regexes.  Maybe something with `r[xyz]` to match rotations?

mtreinish replied:
yeah, or `r*` would probably work too (to include the `RGate` in that set too)

mtreinish commented on releasenotes/notes/0.19/user-config-mpl-style-a9ae4eb7ce072fcd.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    Fixed a small bug where the circuit drawer would ignore the Matplotlib style
+    specified in ``settings.conf``.
```
```suggestion
    Fixed an issue with the :meth:`~qiskit.circuit.QuantumCircuit.draw` method and
    :func:`~qiskit.visualization.circuit_drawer` where a custom style set via the
    user config file (i.e. ``settings.conf``) would ignore the set value of the
    ```circuit_mpl_style`` field if the ``style`` kwarg on the function/method was not
    set.
```

mtreinish commented on releasenotes/notes/0.19/vf2layout-4cea88087c355769.yaml:
```diff
@@ -1,7 +1,7 @@
   - |
-    A new pass :class:`qiskit.transpiler.passes.VF2Layout` is
+    A new transpiler pass :class:`~qiskit.transpiler.passes.VF2Layout` was
     introduced. This pass models the layout allocation problem as a subgraph
     isomorphism problem. For this, it uses VF2, a state of the art heuristic to find
-    subgraphs. The pass works very much like :class:`qiskit.transpiler.passes.CSPLayout`. 
+    subgraphs. The pass works very much like :class:`~qiskit.transpiler.passes.CSPLayout`. 
```
```suggestion
    Added a new transpiler pass, :class:`~qiskit.transpiler.passes.VF2Layout`. This pass models the
    layout allocation problem as a subgraph isomorphism problem and uses the |VF2|_ algorithm implementation
    in `retworkx <https://qiskit.org/documentation/retworkx/stubs/retworkx.vf2_mapping.html>`__ to find a perfect
    layout (a layout which would not require additional routing) if one exists. The functionality exposed by this new
    pass is very similar to exisiting :class:`~qiskit.transpiler.passes.CSPLayout` but
    :class:`~qiskit.transpiler.passes.VF2Layout` is significantly faster.

    .. _VF2: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.101.5342&rep=rep1&type=pdf
```

mtreinish reviewed (commented):
Ok I've finished a first pass now, the only thing missing besides all my inline comments is the prelude section where we should highlight the big features for this release (`BackendV2/Target`, `QASM3`, and control flow). We also should highlight in the prelude that this is the last minor release with python 3.6 support and 0.20 will require python >=3.7

mtreinish commented on qiskit/transpiler/passes/calibration/builders.py:
```diff
@@ -103,7 +103,7 @@ class RZXCalibrationBuilder(CalibrationBuilder):
     """
 
-    def __init__(self, backend: basebackend):
+    def __init__(self, backend: BaseBackend):
```
Resolved with #7331

jakelishman commented on releasenotes/notes/0.19/fix-compute-variance-type-error-2eb47903620b4b08.yaml:
```diff
@@ -0,0 +1,7 @@
+---
+fixes:
+  - |
+    Fixed a ``TypeError`` occurring during :meth:`.PauliExpectation.compute_variance`
```
Yep, looks that way to me. Removed.

jakelishman commented on releasenotes/notes/0.19/fix-decompose-empty-gate-71455847dcaaea26.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Fixed a bug where circuit gates whose definitions were empty circuits were not decomposed.
```
I've put this in, but I'm not sure I agree that we need to do this.  If you'd seen the bug then it'd make sense to you, otherwise it doesn't really tell you all that much, and makes it more daunting to read the release notes, because they go on for screens and screens.  I'm more of a fan of just linking to PRs/issues for bug fixes.

jakelishman commented on releasenotes/notes/0.19/make-statevector-subscriptable-and-add-inner-product_method-a0337393d9a5b666.yaml:
```diff
@@ -1,16 +1,13 @@
+    The :class:`~qiskit.quantum_info.Statevector` class is now subscriptable.
     User can now retrieve the nth element in a :class:`~qiskit.quantum_info.Statevector` by index as ``statevec[n]``.
 
     See `#4916 <https://github.com/Qiskit/qiskit-terra/issues/4916>`__ from more information.
```
I agree with this for features.  I mentioned it before, but I'm not sure I agree with putting significant bugfix details in the release notes (unless it was a major bug), and I definitely just prefer linking issues for those, for the same reasons you give here - I don't think the audience of release notes _usually_ cares about the details of bugfixes, so hiding the detail in a link makes it easier for them to read, while stile giving access to the full context if they want it.

jakelishman commented on releasenotes/notes/0.19/refactor-set-qft-num-qubits-82e6df88448c2f22.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Refactored the :attr:`.QFT.num_qubits` setter to avoid potential problems when changing the number of qubits of the :obj:`.QFT` circuit. 
```
I've actually dropped this release note - in practice there was no bug that could actually have been user-facing, because no matter what the value was, if it were truth-y it would be passed to `QuantumRegister` which would immediately error.  Erroring is the correct thing in this situation.  The old form actually provided a better error message than what we get now, because it does a type check, whereas non-numeric values now fail on `value > 0`.

jakelishman commented on releasenotes/notes/0.19/sparse-pauli-internal-8226b4f57a61b982.yaml:
```diff
@@ -6,7 +6,7 @@ upgrade:
     :class:`~qiskit.quantum_info.operators.PauliList`.
```
I've rewritten this whole release note.  I've removed the upgrade note for `SparsePauliOp`, and expanded the deprecation notice to say more about the getter properties that you should now use.  I split the last line out into its own upgrade note, because that's actually what it is, and changed the warning emitted to a `FutureWarning`, because it's changing behaviour.

I'm a little worried about the speed that `PauliTable` is being completely stripped out of everything, and even deprecated itself.

jakelishman commented on releasenotes/notes/0.19/add-getters-and-setters-for-vqe-edc753591b368980.yaml:
```diff
@@ -0,0 +1,21 @@
+  - |
+    Every attribute of the :class:`~qiskit.algorithms.VQE` class that is set at the initialization is now
+    accessible with getters and setters. Further, the default values of the VQE attributes such as ``ansatz``,
+    ``optimizer`` or ``quantum_instance`` can be reset by setting the attribute to ``None``.
```
Done - it even turned out that `quantum_instance` didn't have the behaviour the note says it does, so I removed that.

jakelishman commented on releasenotes/notes/0.19/add-getters-and-setters-for-vqe-edc753591b368980.yaml:
```diff
@@ -0,0 +1,21 @@
+
+upgrade:
+  - |
+    The ``sort_parameters_by_name`` of the :class:`~qiskit.algorithms.VQE` class has been removed, following its deprecation on July 12th, 2021.
```
The date was in it already (though I think I added the 2021), but it's got the Terra version now too.

jakelishman commented on releasenotes/notes/0.19/pauli-evolution-gate-ad767a3e43714fa7.yaml:
```diff
@@ -0,0 +1,32 @@
+        circuit = QuantumCircuit(3)
+        circuit.append(evo, range(3))
+
+        print(circuit.draw())
```
Yeah, not sure what was going on here.  I just removed the last line.

jakelishman commented:
- Removed release notes from the 0.18 backport series in 3584812e7
- Addressed all the review comments in 2114e3e7e, including reflowing all the text (except for two comments, which I've left unresolved).
- Added a release prelude in fefe0e97b.

mtreinish commented on releasenotes/notes/0.19/refactor-grover-isgoodstate-check-54fdb61f899e5158.yaml:
```diff
@@ -0,0 +1,30 @@
+---
+features:
+  - |
+    If the optimal power of the Grover operator is not known, the :class:`~qiskit.algorithms.Grover` 
+    algorithm checks different powers (i.e. iterations) and applies the 
+    :meth:`~qiskit.algorithms.AmplificationProblem.is_good_state` function to check whether
+    a good bitstring has been measured. However, if the optimal power is known and only a single 
+    circuit is run, the ``is_good_state`` function is not required and Grover search can simply 
+    return the most likely bitstring.
+
+    In this PR, :class:`~qiskit.algorithms.Grover` and 
+    :class:`~qiskit.algorithms.AmplificationProblem` were adapted to support this behavior. For 
+    example, the following would previously throw a ``TypeError`` but is now valid:
```
I'm not sure if this is still clear, but I think the wording needs to be adjusted a bit to make it sound like a release note.

```suggestion
    When running the :class:`~qiskit.algorithms.Grover`algorithm class if the optimal power is
    known and only a single circuit is run, the :meth:`.AmplificationProblem.is_good_state` method
    is no longer required to be set and the Grover search will return the most likely bitstring.

    Generally, if the optimal power of the Grover operator is not known, the :class:`~qiskit.algorithms.Grover` 
    algorithm checks different powers (i.e. iterations) and applies the 
    :meth:`~qiskit.algorithms.AmplificationProblem.is_good_state` function to check whether
    a good bitstring has been measured.  For example, you are now able to run something like:
```

jakelishman replied:
Done in 905141d

Cryoris commented on releasenotes/notes/0.19/qaoa-parameters-49e4524ed2d3e875.yaml:
```diff
@@ -0,0 +1,9 @@
+    :class:`~qiskit.circuit.library.QAOAAnsatz` were not been set.
+other:
+  - |
+    Terra now uses the QAOA notation for the cost operator (``γ``) and mixer (``β``) operator parameters in
+    :class:`~qiskit.circuit.library.QAOAAnsatz`.
```
How about:
```suggestion
    The :class:`~qiskit.circuit.library.QAOAAnsatz` has been updated to use the parameter
    symbol ``γ`` for the cost operator and ``β`` for the mixer operator, as is the standard
    notation in QAOA literature.
```

mtreinish commented on releasenotes/notes/0.19/0.19-prelude-65c295aa9497ed48.yaml:
```diff
@@ -0,0 +1,28 @@
+      :class:`~.transpiler.Target` to allow backends to better model their
+      constraints for the :ref:`transpiler <qiskit-transpiler>`.
+
+    * An :ref:`extensible plugin interface <qiskit-transpiler-plugins>` to the
```
Oh good call adding this here, I didn't think of that but it's a good one to highlight too.

mtreinish reviewed (commented):
Thanks for updating everything after all my comments before. I think this is almost ready (once the last 2 high priority 0.19 PRs merge). Just a couple of small inline comments

jakelishman commented on releasenotes/notes/0.19/remove-manual-warning-filters-028646b73bb86860.yaml:
```diff
@@ -0,0 +1,19 @@
+    blamed with a warning if internal Qiskit functions call deprecated
+    behaviour.  If you write libraries, you should occasionally run with the
+    default warning filters disabled, or have tests which always run with them
+    disabled.
```
Done in 905141d

nkanazawa1989 reviewed (commented):
Pulse documentation upgrade looks nice. It's much cleaner than before.

1ucian0 commented on qiskit/pulse/builder.py:
```diff
@@ -10,13 +10,48 @@
+The pulse builder provides an imperative API for writing pulse programs
+with less difficulty than the :class:`~qiskit.pulse.Schedule` API.
+It contextually constructs a pulse schedule and then emits the schedule for
+execution. For example to play a series of pulses on channels is as simple as:
```
```suggestion
execution. For example, to play a series of pulses on channels is as simple as:
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/refactor-grover-isgoodstate-check-54fdb61f899e5158.yaml:
```diff
@@ -0,0 +1,29 @@
+
+      # Run Grover search and print the best measurement
+      result = grover.amplify(problem)
+      print(result.top_measurement)  # should print 111
```
This is wrongly rendered. 
<img width="606" alt="Screenshot 2021-12-03 at 15 49 20" src="https://user-images.githubusercontent.com/766693/144622617-22765119-f37c-4e2f-a4ae-1b2da326c289.png">



It might be the lack of trailing space?

```suggestion
      print(result.top_measurement)  # should print 111

```

jakelishman replied:
This is because I've accidentally used two different syntax forms for introducing a code block: the paragraph ends in `::`, but then it starts `.. code-block::` as well.  I think it may have been a merge mistake on my part - I think I updated the note locally, and then copied in some suggestion from Matthew or Abby without paying attention.

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/fix-mpl-drawer-bit-condition-90c4dac2defdd8c6.yaml:
```diff
@@ -0,0 +1,7 @@
+    The ``mpl`` output method for the :func:`~qiskit.visualization.circuit_drawer`
+    function and the :meth:`.QuantumCircuit.draw` method can now 
+    draw circuits that contain gates with single bit condition. This was added for
+    compatibility of mpl drawer with the new feature of supporting classical 
```
```suggestion
    compatibility of ``'mpl '``drawer with the new feature of supporting classical 
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/qasm3_dumps-7475de655e1acb24.yaml:
```diff
@@ -14,9 +15,10 @@ features:
-    This initial release has limited support for named registers, basic builtin instructions
-    (just as measure and barriers), user-defined gates, user-defined instructions (as
-    subroutines), and opaque gates/instructions as ``defcal``.
+    This initial release has limited support for named registers, basic builtin
```
```suggestion
    This initial release has limited support for named registers, basic built-in
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/qdrift-as-product-formula-044a37a106a45a47.yaml:
```diff
@@ -1,19 +1,16 @@
+    synthesis method for :obj:`.PauliEvolutionGate`, deriving from
+    :obj:`~qiskit.opflow.evolutions.TrotterizationBase`.
 
     .. code-block:: python
```
Because this finishes with a print, I assume the intention was to execute it.

```suggestion
    .. jupyter-execute::
```

jakelishman replied:
I've been moving away from `jupyter-execute` in the release notes a little bit, because by definition release notes go out-of-date but stay in the documentation, so they'll often stale.  I probably changed the directive and forgot to remove the `print`.  The important bit here is the creation of the `PauliEvolutionGate`.

jakelishman replied:
Done in e624bd6.

1ucian0 replied:
>  [...] by definition release notes go out-of-date but stay in the documentation, so they'll often stale.
 
This is a really good point. I think is reasonable to stop using execute in release note as a policy.

1ucian0 commented on releasenotes/notes/0.19/quantumcircuit-dynamic-instructions-a69db25665739004.yaml:
```diff
@@ -2,19 +2,19 @@
     context, respectively.
 
-    These can be created as stand alone :class:`~qiskit.circuit.Instruction` s,
+    These can be created as stand alone :class:`~qiskit.circuit.Instruction`\ s,
```
```suggestion
    These can be created as stand-alone :class:`~qiskit.circuit.Instruction`\ s,
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/retworkx-substitute_node_with_dag-speedup-d7d1f0d33716131d.yaml:
```diff
@@ -4,6 +4,6 @@ features:
     circuit representation.  This can often result in speedups in calls to
-    :obj:`~qiskit.transpile` of around 10-40%, with greater effects at higher
+    :obj:`~.compiler.transpile` of around 10-40%, with greater effects at higher
     optimisation levels.  See `#6302
```
```suggestion
    optimization levels.  See `#6302
```

jakelishman replied:
It shows that I'm the one who edited these, huh?

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/execute-fix-108e835dc4f4593d.yaml:
```diff
@@ -1,13 +1,13 @@
     Previously these defaults were set to match the default values when
-    calling :func:`qiskit.execute_function.execute` with a legacy
+    calling :func:`~qiskit.execute_function.execute` with a legacy
     :class:`~qiskit.providers.BaseBackend` backend. For example running::
```
```suggestion
    :class:`~qiskit.providers.BaseBackend` backend. For example ::
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/symengine-bump-20dedd5204870e7a.yaml:
```diff
@@ -1,8 +1,8 @@
   - |
-    The minimum version of syengine required for installing has been increased
+    The minimum version of Symengine required for installing has been increased
     to 0.8.0. This was neccessary to fix some issues with the handling of
```
```suggestion
    to 0.8.0. This was necessary to fix some issues with the handling of
```

jakelishman replied:
Done in e624bd6.

1ucian0 commented on releasenotes/notes/0.19/QuantumCircuit.decompose-takes-which-gate-to-decompose-d857da5d0c41fb66.yaml:
```diff
@@ -0,0 +1,24 @@
+    :class:`~qiskit.transpiler.passes.Decompose` now optionally accept a
+    parameter containing a collection of gate names.  If this parameter is given,
+    then only gates with matching names will be decomposed.  This supports
+    Unix-shell-style wildcard matches.  For example::
+
+        qc.decompose(["h", "r[xz]"])
+
+    will decompose any ``h``, ``rx`` or ``rz`` gates, but leave (for example) ``x`` gates untouched.
```
Spacing after a code block is big and breaks the sentence into a paragraph:
<img width="655" alt="Screenshot 2021-12-04 at 07 51 14" src="https://user-images.githubusercontent.com/766693/144700686-afa15c65-0b25-46ba-815d-f47d7bc258fa.png">



```suggestion
    Unix-shell-style wildcard matches.  For example, ``qc.decompose(["h", "r[xz]"])`` will decompose any ``h``, ``rx`` or ``rz`` gates, but leave (for example) ``x`` gates untouched.
```

jakelishman replied:
This particular one is a CSS problem, not a release note problem, I think - if you look at places where there's a paragraph break with _no_ code, there's no space between paragraphs (but there probably should be), whereas here there's a big space, but it shouldn't be any more than a regular paragraph space.

A quick look shows it's coming from an `article.pytorch-article pre { margin-bottom: 2.5rem; }` CSS rule; I'm not sure we should have that at all, but if it's useful elsewhere, we could also override it with something like a 
```css
#release-notes article.pytorch-article pre {
    margin-bottom: 0;
}
```
because really we only want the paragraph spacing.

jakelishman replied:
We definitely want this rule - it's the same thing that causes a weirdly large space in the text-only output from a `jupyter-execute` block.

1ucian0 commented on releasenotes/notes/0.19/add-passmanager-config-from-backend-af5dd7d99ec053ef.yaml:
```diff
@@ -0,0 +1,16 @@
+    Added a new constructor method :meth:`.PassManagerConfig.from_backend`.  It
+    constructs a :class:`~qiskit.transpiler.PassManagerConfig` object with user
+    options and the configuration of a backend.  With this feature, a preset
+    passmanager can be built easier, for example::
```
```suggestion
    passmanager can be built easier. For example::
```

jakelishman replied:
Done in 1423ebf.

1ucian0 commented on releasenotes/notes/0.19/full_prec_sympy-aeee8210091ef20f.yaml:
```diff
@@ -0,0 +1,14 @@
+    parameters that are bound to floats.  This has consequences for QASM
+    serialization and the circuit text drawer::
+
+    .. code-block:: python
+
+      >>> from qiskit.circuit import Parameter
+      >>> x = Parameter('x')
+      >>> str(x.bind({x:0.5}))
+      '0.5'   # instead of '0.500000000000000'
```
<img width="734" alt="Screenshot 2021-12-04 at 08 22 49" src="https://user-images.githubusercontent.com/766693/144701466-95ee9998-fd00-4925-840c-4d31b4a81c8a.png">



```suggestion
    .. code-block:: python

        >>> from qiskit.circuit import Parameter
        >>> x = Parameter('x')
        >>> str(x.bind({x:0.5}))
        '0.5'   # instead of '0.500000000000000'
```

jakelishman replied:
Done in 1423ebf.

1ucian0 commented on releasenotes/notes/0.19/user-config-mpl-style-a9ae4eb7ce072fcd.yaml:
````diff
@@ -0,0 +1,8 @@
+    Fixed an issue with the :meth:`~qiskit.circuit.QuantumCircuit.draw` method and
+    :func:`~qiskit.visualization.circuit_drawer` function, where a custom style set via the
+    user config file (i.e. ``settings.conf``) would ignore the set value of the
+    ```circuit_mpl_style`` field if the ``style`` kwarg on the function/method was not
````
```suggestion
    ``circuit_mpl_style`` field if the ``style`` kwarg on the function/method was not
```

jakelishman replied:
Done in 1423ebf.

1ucian0 commented:
I dont thinks is possible to suggest changes in unmodified moved files. So Ill drop them here:
 * `add-backend-v2-ce84c976fb13b038.yaml:33`:     rest of Qiskit can use today with any backend. Subclasses of the abstract

1ucian0 commented on releasenotes/notes/0.19/remove-schemas-ca9f3f2e0f08bca8.yaml:
```diff
@@ -4,33 +4,32 @@ upgrade:
-    The exception raised by the :func:`qiskit.compiler.assemble` function when
+    The exception raised by the :func:`~.compiler.assemble` function when
     invalid parameters are passed in for constructing a
     :class:`~qiskit.qobj.PulseQobj` have changed from a ``SchemaValidationError`
```
```suggestion
    :class:`~qiskit.qobj.PulseQobj` have changed from a ``SchemaValidationError``
```

jakelishman replied:
Done in 1423ebf.

1ucian0 commented on releasenotes/notes/0.19/taper_empty_operator_fix-53ce20e5d2b68fd6.yaml:
````diff
@@ -1,11 +1,11 @@
     The code will now correctly taper a zero operator such that the number of qubits is
-    reduced as expected and matches to tapered non-zero operators e.g 0*"IIII" when we are
-    tapering by 3 qubits will become 0*"I".
+    reduced as expected and matches to tapered non-zero operators e.g ```0*"IIII"``` when we are
+    tapering by 3 qubits will become ```0*"I"```.
````
```suggestion
    reduced as expected and matches to tapered non-zero operators e.g ``0*"IIII"`` when we are
    tapering by 3 qubits will become ``0*"I"``.
```

jakelishman replied:
Done in 1423ebf.

jakelishman commented:
Resolved all Luciano's most recent comments in 1423ebf.  The styling issue I think needs fixing in our theme.

mtreinish reviewed (commented):
Almost done with my final pass (need to step away for a bit right now) so I'll likely have some other comments when I get back

eggerdj commented on releasenotes/notes/0.19/add-pulse-gate-pass-dc347177ed541bcc.yaml:
```diff
@@ -0,0 +1,100 @@
+        from qiskit.transpiler.passes.scheduling.calibration_creators import RZXCalibrationBuilder
+        from qiskit.transpiler.passes.scheduling.calibration_creators import RZXCalibrationBuilderNoEcho
+
+    are deprecated.
```
Perhaps add a statement regarding where these should be imported from?

jakelishman replied:
Done in a40c87764.

eggerdj commented on releasenotes/notes/0.19/add-pulse-gate-pass-dc347177ed541bcc.yaml:
```diff
@@ -0,0 +1,100 @@
+    Any transpiler pass that interacts with the pulse gate feature will be
+    grouped to ``qiskit.transpiler.passes.calibration``.
+
+    This pass is applied to all optimization levels from 0 to 3.
+    No gate implementation is updated unless the end-user explicitly overrides
+    the ``backend.defaults().instruction_schedule_map``.
+
+    This pass saves users from individually calling
+    :meth:`.QuantumCircuit.add_calibration` for every circuit run on the
+    hardware.
```
It is `PulseGates`. See here: https://github.com/Qiskit/qiskit-terra/blob/1a9f93804f2696e3a6c119ecbebd3724df008dfa/qiskit/transpiler/preset_passmanagers/level0.py#L262

mtreinish replied:
We should call that out explicitly then, or realistically do that and make this a separate bullet point because it's really a separate release note than the first part here.

jakelishman replied:
I had a bit of a rewrite in a40c877, but I may have got some of that not 100% correct. Would be good if someone with more knowledge could check.

eggerdj commented on releasenotes/notes/0.19/add-pulse-gate-pass-dc347177ed541bcc.yaml:
```diff
@@ -0,0 +1,100 @@
+    :meth:`.QuantumCircuit.add_calibration` for every circuit run on the
+    hardware.
+
+    Along with this change, a schedule was added to
+    :class:`~qiskit.pulse.InstructionScheduleMap` and is implicitly updated with
+    a metadata ``"publisher"``.  Backend-calibrated gate schedules have a
+    special publisher kind to avoid overriding circuits with calibrations of
+    already known schedules.  Usually, end-users don't need to take care of this
+    metadata as it is applied automatically.  You can call
+    :meth:`.InstructionScheduleMap.has_custom_gate` to check if the map has
+    custom gate calibration.
```
@nkanazawa1989

jakelishman replied:
I had a bit of a rewrite in a40c87764, but I may have got some of that not 100% correct.  Would be good if someone with more knowledge could check.

jakelishman commented on releasenotes/notes/0.19/add-pulse-gate-pass-dc347177ed541bcc.yaml:
```diff
@@ -0,0 +1,100 @@
+    whereas :obj:`.PulseGates` is the new pass that automatically extracts
+    user-provided calibrations from the instruction schedule map and attaches
+    the gate schedule to the given (transpiled) quantum circuit as a pulse gate.
+    Any transpiler pass that interacts with the pulse gate feature will be
+    grouped to ``qiskit.transpiler.passes.calibration``.
```
I'm not sure about a lot of this note - the grouping of passes into `scheduling` or `calibration` is pretty meaningless from an end-user perspective if they're importing the way we suggest they do, with
```python
from qiskit.transpiler.passes import ...
```
because the group doesn't appear.

mtreinish replied:
Yeah we can just remove this piece as not user facing I think. We have the deprecation about the import path change below if people are using the full path for import so I think we don't really need to mention this as a new feature.

jakelishman replied:
Removed in a40c87764.

jakelishman commented on releasenotes/notes/0.19/add-pulse-gate-pass-dc347177ed541bcc.yaml:
```diff
@@ -0,0 +1,100 @@
+
+    - :class:`~qiskit.transpiler.passes.RZXCalibrationBuilder`
+    - :class:`~qiskit.transpiler.passes.RZXCalibrationBuilderNoEcho`
+    - :func:`~qiskit.transpiler.passes.calibration.rzx_templates.rzx_templates`.
```
Even looking at the function, it seems quite unrelated to calibrations, so I'm not 100% sure why it's there.  We can remove the link, for sure.

jakelishman replied:
Removed this as part of a40c87764.

jakelishman commented on releasenotes/notes/0.19/draw-statevector-in-ket-notation-0726959d1f6ea3ce.yaml:
```diff
@@ -0,0 +1,16 @@
+
+        sv = Statevector([math.sqrt(0.5), 0, 0, -math.sqrt(0.5)])
+        sv.draw(output='latex')
+        sv.draw(output='latex', convention='vector')  # old behaviour
```
Yep, looks like it changed the default, but only for `Statevector.draw("latex")`.  The default behaviour is (and was) basically a bug - it just prints out the `repr`, and doesn't do any drawing.

jakelishman replied:
Actually, I'm going to revert to the previous default behaviour for `Statevector.draw("latex")` here - the "ket" form is broken with anything other than vectors of qubits:

<img width="1063" alt="Screenshot 2021-12-06 at 13 38 55" src="https://user-images.githubusercontent.com/5968590/144855952-7e48ebc7-da5e-448b-b75a-e86eeabb115b.png">

mtreinish replied:
Yeah lets just change the default back to `vector` quickly and maybe document that as a known issue here

jakelishman replied:
This actually turned out to be a deliberate choice (#6154).  I've raised a bug (#7355) against the behaviour, but I guess we should keep it for now.  I've added an upgrade note and a bug note in 33ffd86f4, and hopefully it won't bite too many people anyway.

mtreinish commented on releasenotes/notes/0.19/readout-mitigation-classes-2ef175e232d791ae.yaml:
```diff
@@ -1,55 +1,55 @@
-    property and init kwarg to :meth:`qiskit.result.QuasiDistribution` for
-    storing standard error in quasi probability estimates. This is used by
+    Added a :attr:`.QuasiDistribution.stddev_upper_bound`
+    attribute and init kwarg to :meth:`qiskit.result.QuasiDistribution` for
```
```suggestion
    attribute and a kwarg to the constructor for the :class:`qiskit.result.QuasiDistribution` class which is used for
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/7274-6f57628a7995a461.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+upgrade:
+  - |
+    In this version, ``from qiskit import *`` will not import submodules, but
```
This just reads like it's only for 0.19.0, maybe something like:

```suggestion
    Starting with this version, ``from qiskit import *`` will not import submodules, but
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/sparse-pauli-internal-8226b4f57a61b982.yaml:
```diff
@@ -0,0 +1,15 @@
+upgrade:
+  - |
+    The return type of :func:`~qiskit.quantum_info.pauli_basis` will change from
+    `~qiskit.quantum_info.PauliTable` to `~qiskit.quantum_info.PauliList` in a
```
```suggestion
    :class:`~qiskit.quantum_info.PauliTable` to :class:`~qiskit.quantum_info.PauliList` in a
```

mtreinish commented on releasenotes/notes/0.19/symengine-bump-20dedd5204870e7a.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+upgrade:
+  - |
+    The minimum version of Symengine required for installing has been increased
```
We probably should add a link to it on pypi:
```suggestion
    The minimum version of `Symengine <https://pypi.org/project/symengine/>`__ required for
    installing has been increased
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/QuantumCircuit.decompose-takes-which-gate-to-decompose-d857da5d0c41fb66.yaml:
```diff
@@ -0,0 +1,24 @@
+deprecations:
+  - |
+    The ``gate`` attribute and initialization parameter of
+    :obj:`qiskit.transpiler.passes.basis.Decompose` is deprecated, and will be
```
```suggestion
    :obj:`qiskit.transpiler.passes.Decompose` is deprecated, and will be
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/dag_node_to_op_in_out_node-af2cf9c3e7686285.yaml:
```diff
@@ -0,0 +1,18 @@
+---
+deprecations:
+  - |
+    The :class:`~qiskit.dagcircuit.dagnode.DAGNode` is being deprecated as a
+    standalone class and will be used in the future only as the parent class for
+    :class:`~qiskit.dagcircuit.dagnode.DAGOpNode`,
+    :class:`~qiskit.dagcircuit.dagnode.DAGInNode`, and
+    :class:`~qiskit.dagcircuit.dagnode.DAGOutNode`. As part of this deprecation, the
+    following kwargs and associated attributes in :obj:`.DAGNode` are also being
+    deprecated: ``type``, ``op``, and ``wire``.
```
```suggestion
    The :class:`~qiskit.dagcircuit.DAGNode` class is being deprecated as a
    standalone class and will be used in the future only as the parent class for
    :class:`~qiskit.dagcircuit.DAGOpNode`,
    :class:`~qiskit.dagcircuit.DAGInNode`, and
    :class:`~qiskit.dagcircuit.DAGOutNode`. As part of this deprecation, the
    following kwargs and associated attributes in :obj:`.DAGNode` are also being
    deprecated: ``type``, ``op``, and ``wire``.
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/fix-configurable-fake-backend-6a07ca5a6159baf5.yaml:
```diff
@@ -2,11 +2,11 @@
-    access the ``backend_name`` attribute instead.
-    in several unittests because string is not callable.
+    ``name`` attribute directly before you will now need to either call it as a
+    method or access the ``backend_name`` attribute instead.  in several
+    unittests because string is not callable.
```
```suggestion
    method or access the ``backend_name`` attribute instead.
```

jakelishman replied:
Dunno what happened here!  Fixed in 33ffd86f4.

mtreinish commented on releasenotes/notes/0.19/fix-transpile-empty-list-c93a41d4145a01c3.yaml:
```diff
@@ -0,0 +1,6 @@
+fixes:
+  - |
+    Calling :obj:`~qiskit.compiler.transpile` on an empty list will now
+    correctly return an empty list without issuing a warning.  Fix of `#7287
```
```suggestion
    correctly return an empty list without issuing a warning.  Fixed `#7287
```

jakelishman replied:
Done in 33ffd86f4.

mtreinish reviewed (commented):
I quickly skimmed through the fix notes and caught a few small things but for the most part it looked okay. I think once the issues I commented on before and these are fixed we should be good to go on this.

mtreinish commented:
I've removed the old hold here because all the release PRs are merged. Once we're done with the release notes this is ready to merge and we can tag the release.

jakelishman commented on releasenotes/notes/0.19/0.19-prelude-65c295aa9497ed48.yaml:
```diff
@@ -0,0 +1,28 @@
+
+    .. _OpenQASM 3 specification: https://qiskit.github.io/openqasm/
+
+    As always, there are many more features and fixes in this release as well,
+    which you can read about below.
```
Done in 33ffd86f4.

jakelishman commented on releasenotes/notes/0.19/QuantumCircuit.decompose-takes-which-gate-to-decompose-d857da5d0c41fb66.yaml:
```diff
@@ -0,0 +1,24 @@
+
+    will decompose any ``h``, ``rx`` or ``rz`` gates, but leave (for example) ``x`` gates untouched.
+
+    See  `#2906 <https://github.com/Qiskit/qiskit-terra/issues/2906>`__ for more information.
```
Nope, removed in 33ffd86f4.

jakelishman commented on releasenotes/notes/0.19/SPSA-termination-callback-a1ec14892f553982.yaml:
```diff
@@ -1,7 +1,7 @@
 features:
   - |
-    Add `termination_callback` argument to :class:`qiskit.algorithms.optimizers.spsa.SPSA` optimizer.
+    Added the ``termination_callback`` argument to the :class:`~qiskit.algorithms.optimizers.SPSA` optimizer.
```
Well spotted - done in 33ffd86f4.

jakelishman commented on releasenotes/notes/0.19/add-detach-prefix-088e96b88ba29927.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Added a utility function :func:`~qiskit.utils.detach_prefix` that is a
```
I generally assume people will click on the link if they're not sure how to access something, but I dropped the tilde in 33ffd86f4.

jakelishman commented on releasenotes/notes/0.19/readout-mitigation-classes-2ef175e232d791ae.yaml:
```diff
@@ -1,55 +1,55 @@
     :class:`~qiskit.utils.QuantumInstance` class still use the legacy mitigators
-    migrated from qiskit-ignis in :mod:`qiskit.utils.mitigation`. It is planned
+    migrated from Qiskit Ignis in :mod:`qiskit.utils.mitigation`. It is planned
     to upgrade code to use the new mitigator classes and deprecate the legacy
```
Done in 33ffd86f4.

jakelishman commented:
All changes now addressed (including removing the gate type hints note).  I rewrote the `PulseGate` note a little bit (conversation still unresolved above), so it'd be good for someone with more knowledge to check that.

mtreinish reviewed (dismissed):
LGTM, it looks like you addressed all the comments thanks for the quick turnaround.

For the pulse gate release note if it's not clear we can always update it after the release. I think at this point it's good enough in it's current form.
